File: php/src/viber.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function viber_parse_event(array $payload): array
{
    $sender = $payload['sender'] ?? []; $message = $payload['message'] ?? []; $user = $sender['id'] ?? ''; if ($user === '') throw new InvalidArgumentException('Viber event has no sender'); $type = $message['type'] ?? 'text'; return ['platform'=>'viber','user_id'=>(string)$user,'content_type'=>$type,'text'=>$message['text'] ?? null,'media_url'=>$message['media'] ?? null];
}
function viber_send_reply(array $event, array $reply, string $token): void
{
    if ($token === '') throw new InvalidArgumentException('VIBER_AUTH_TOKEN is required'); foreach (array_slice($reply['messages'] ?? [],0,5) as $message) { $type=$message['type'] ?? 'text'; if ($type==='audio') throw new InvalidArgumentException('Viber audio replies are not supported'); $body=['receiver'=>$event['user_id'],'type'=>$type]; if($type==='text')$body['text']=$message['text']??'';else{$body['media']=$message['media_url']??'';if($body['media']==='')throw new InvalidArgumentException('Viber media reply requires URL');if(!empty($message['text']))$body['text']=$message['text'];}$ch=curl_init('https://chatapi.viber.com/pa/send_message');curl_setopt_array($ch,[CURLOPT_POST=>true,CURLOPT_RETURNTRANSFER=>true,CURLOPT_HTTPHEADER=>['X-Viber-Auth-Token: '.$token,'Content-Type: application/json'],CURLOPT_POSTFIELDS=>json_encode($body,JSON_THROW_ON_ERROR)]);$result=curl_exec($ch);$status=curl_getinfo($ch,CURLINFO_RESPONSE_CODE);if($result===false||$status<200||$status>=300)throw new RuntimeException('Viber API request failed');curl_close($ch);}
}

File: python/platforms/viber.py
from __future__ import annotations
import os
from typing import Any
import requests
from core.models import InboundEvent, OutboundReply
from .base import PlatformAdapter
from .catalog import PLATFORM_CATALOG

class ViberAdapter(PlatformAdapter):
    def __init__(self, token: str | None = None, session: requests.Session | None = None):
        self.capabilities = PLATFORM_CATALOG["viber"]; self.token = token or os.getenv("VIBER_AUTH_TOKEN")
        if not self.token: raise ValueError("VIBER_AUTH_TOKEN is required")
        self.session = session or requests.Session()
    def parse_event(self, payload: Any, headers=None) -> InboundEvent:
        sender = payload.get("sender", {}); message = payload.get("message", {}); kind = message.get("type", "text")
        kind = kind if kind in ("text", "image", "video", "file", "audio") else "text"
        return InboundEvent(platform="viber", user_id=str(sender.get("id") or payload.get("user_id")), content_type=kind, text=message.get("text"), media_url=message.get("media"))
    def send_reply(self, event: InboundEvent, reply: OutboundReply) -> None:
        for message in reply.messages:
            body = {"receiver": event.user_id, "type": message.type}
            if message.type == "text":
                body["text"] = message.text
            else:
                if message.type == "audio":
                    raise ValueError("Viber audio replies are not supported by the documented Bot API types")
                if not message.media_url:
                    raise ValueError(f"Viber {message.type} reply requires media_url")
                body["media"] = message.media_url
                if message.text:
                    body["text"] = message.text
            response = self.session.post("https://chatapi.viber.com/pa/send_message", headers={"X-Viber-Auth-Token": self.token}, json=body, timeout=30); response.raise_for_status()
